### PR TITLE
Reader+Writer Integration Tests

### PR DESCRIPTION
**Pull request by puerco**
This PR adds the integration tests for the reader and writer packages.

To properly fake the implementations, the serializer and unserializer interfaces and implementations have now moved into a new `native` package. This breaks the import loop from the reader/writer packages and its fakes.

The PR also adds the autogenerated fakes and a target in the main makefile to regenerate them. It also adds a presubmit to ensure the fakes are always up to date.

Signed-off-by: Adolfo García Veytia (Puerco) <puerco@chainguard.dev>

### CODE REVIEW

File: pkg/writer/writer_test.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+type fakeWriteCloser struct {
+	*bufio.Writer
+}
+
+func (fwc *fakeWriteCloser) Close() error {
+	return nil
+}
+
+func TestWriteStream(t *testing.T) {
```
Comment (jspeed-meyers) [inline pkg/writer/writer_test.go:23]: Also great!

**Review by jspeed-meyers — APPROVED** [review]
This is excellent. We can always add more tests if others want more test :)

Nice!

File: .github/workflows/go-build-and-test.yml [hunk 2]
```diff
@@ -16,6 +16,10 @@ jobs:
       with:
         go-version: '1.20'
 
+    - name: Verify autogenerated fakes
+      run: |
+        hack/verify-fakes.sh
```
[the remark this answers is not in the document]
Comment (puerco) [reply .github/workflows/go-build-and-test.yml:21]: Ah sorry for the lack of context! The `hack` directory.. it is [an old Kubernetes tradition](https://github.com/kubernetes/kubernetes/tree/master/hack) where we store this sort of scripts. I'm happy to rename it to something else.

(Fun fact, it was invented by a coworker of ours 🙂)

[the remark this answers is not in the document]
Comment (jspeed-meyers) [reply .github/workflows/go-build-and-test.yml:21]: I thought that coworker might be indirectly responsible :)

Okay, makes sense. Maybe add a quick README with a little explanation and pointer?

[the remark this answers is not in the document]
Comment (puerco) [reply .github/workflows/go-build-and-test.yml:21]: Done https://github.com/bom-squad/protobom/pull/87

**Pull request merged**